File: ios/Shared/Boundary/BoundaryService.swift
import Foundation

enum BoundaryServiceError: Error { case nullResult, startFailed(Int) }

/// The single seam the Boundary screen talks to; `FFIBoundaryService` is the real C-ABI impl,
/// `MockBoundaryService` backs tests and previews.
protocol BoundaryService: Sendable {
    func echo(_ input: String, skipFree: Bool) async throws -> BoundaryEchoTrace
    func throwDemo() async throws -> BoundaryThrow
    func stream(_ prompt: String, maxTokens: Int) -> AsyncThrowingStream<BoundaryStreamToken, Error>
}

/// Microseconds between two ContinuousClock instants (ContinuousClock: iOS 16+;
/// pre-iOS 16 alternative: DispatchTime.now().uptimeNanoseconds deltas / 1000).
private func microseconds(_ a: ContinuousClock.Instant, _ b: ContinuousClock.Instant) -> Double {
    let d = a.duration(to: b)
    return Double(d.components.seconds) * 1_000_000 + Double(d.components.attoseconds) / 1e12
}

/// BoundaryService over the in-process C ABI (Pattern 3 — boundary instrumentation).
struct FFIBoundaryService: BoundaryService {

    // MARK: echo / pixels — synchronous, byte + timing inspector
    func echo(_ input: String, skipFree: Bool) async throws -> BoundaryEchoTrace {
        try await Task.detached(priority: .userInitiated) { () throws -> BoundaryEchoTrace in
            var tid: UInt64 = 0
            pthread_threadid_np(nil, &tid)
            let clock = ContinuousClock()

            let mStart = clock.now
            let bytes = Array(input.utf8)
            let mEnd = clock.now

            let cStart = clock.now
            let ptr: UnsafePointer<CChar>? = bytes.withUnsafeBytes { raw in
                dni_ffi_echo(raw.baseAddress?.assumingMemoryBound(to: CChar.self), Int32(bytes.count))
            }
            let cEnd = clock.now
            guard let ptr else { throw BoundaryServiceError.nullResult }
            let json = String(cString: ptr)

            let fStart = clock.now
            // Leak demo: the CALLER owns the free. Skipping it is what makes the outstanding-bytes counter climb.
            if !skipFree { dni_string_free(ptr) }
            let fEnd = clock.now

            let echo = try JSONDecoder().decode(BoundaryEcho.self, from: Data(json.utf8))
            let timing = PhaseTiming(
                marshalUs: microseconds(mStart, mEnd),
                crossUs: max(0, microseconds(cStart, cEnd) - echo.executeUs), // round-trip minus native execute
                executeUs: echo.executeUs,
                callbackUs: 0,
                freeUs: microseconds(fStart, fEnd))
            return BoundaryEchoTrace(echo: echo, timing: timing, callerThreadId: tid, leakedFree: skipFree)
        }.value
    }

    // MARK: throw — contained managed exception
    func throwDemo() async throws -> BoundaryThrow {
        try await Task.detached(priority: .userInitiated) { () throws -> BoundaryThrow in
            guard let ptr = dni_ffi_throw() else { throw BoundaryServiceError.nullResult }
            defer { dni_string_free(ptr) }
            return try JSONDecoder().decode(BoundaryThrow.self, from: Data(String(cString: ptr).utf8))
        }.value
    }

    // MARK: stream — the off-UI-thread callback hop
    private final class StreamBox {
        let continuation: AsyncThrowingStream<BoundaryStreamToken, Error>.Continuation
        init(_ c: AsyncThrowingStream<BoundaryStreamToken, Error>.Continuation) { self.continuation = c }
    }

    func stream(_ prompt: String, maxTokens: Int) -> AsyncThrowingStream<BoundaryStreamToken, Error> {
        AsyncThrowingStream { continuation in
            let box = StreamBox(continuation)
            let userData = Unmanaged.passRetained(box).toOpaque()
            let userDataBits = UInt(bitPattern: userData) // Sendable bit-pattern for the @Sendable onTermination

            // dni_trace_cb: like dni_token_cb but with (int64 managedThreadId, int64 elapsedUs) appended.
            // @convention(c) closure capturing nothing: required to pass as a C function pointer (Swift 5+).
            let callback: @convention(c) (UnsafeMutableRawPointer?, Int32, UnsafePointer<CChar>?, Int32, Int64, Int64) -> Void = { ud, index, text, isFinal, threadId, elapsedUs in
                guard let ud else { return }
                let box = Unmanaged<StreamBox>.fromOpaque(ud).takeUnretainedValue()
                let token = BoundaryStreamToken(
                    index: Int(index),
                    text: text.map { String(cString: $0) } ?? "",
                    isFinal: isFinal != 0,
                    managedThreadId: Int(threadId),
                    elapsedUs: Int(elapsedUs))
                box.continuation.yield(token)
                if isFinal != 0 { box.continuation.finish() }
            }

            let sessionId = prompt.withCString { p in
                dni_ffi_stream_start(p, Int32(maxTokens), callback, userData)
            }
            guard sessionId > 0 else {
                continuation.finish(throwing: BoundaryServiceError.startFailed(Int(sessionId)))
                Unmanaged<StreamBox>.fromOpaque(userData).release()
                return
            }
            // Cancel+free OUTSIDE the callback: dni_session_free blocks on the .NET drain thread the callback
            // runs on, so freeing there would deadlock (same rule as FFIRagService).
            continuation.onTermination = { _ in
                _ = dni_session_cancel(sessionId)
                _ = dni_session_free(sessionId)
                if let p = UnsafeMutableRawPointer(bitPattern: userDataBits) {
                    Unmanaged<StreamBox>.fromOpaque(p).release()
                }
            }
        }
    }
}

/// Deterministic mock for tests/previews — no FFI, no device needed.
struct MockBoundaryService: BoundaryService {
    func echo(_ input: String, skipFree: Bool) async throws -> BoundaryEchoTrace {
        let hex = input.utf8.map { String(format: "%02X", $0) }.joined()
        let echo = BoundaryEcho(bytesHex: hex, len: input.utf8.count, decoded: input,
                                managedThreadId: 7, executeUs: 4.2, ptrIn: "0x1000")
        let timing = PhaseTiming(marshalUs: 1.1, crossUs: 2.0, executeUs: 4.2, callbackUs: 0, freeUs: 0.5)
        return BoundaryEchoTrace(echo: echo, timing: timing, callerThreadId: 1, leakedFree: skipFree)
    }
    func throwDemo() async throws -> BoundaryThrow {
        BoundaryThrow(caught: true, type: "System.InvalidOperationException",
                      message: "Boundary demo: managed exception crossing prevented.", status: -5)
    }
    func stream(_ prompt: String, maxTokens: Int) -> AsyncThrowingStream<BoundaryStreamToken, Error> {
        AsyncThrowingStream { c in
            Task {
                for i in 0..<3 {
                    c.yield(BoundaryStreamToken(index: i, text: "tok\(i) ", isFinal: false, managedThreadId: 9, elapsedUs: i * 800))
                }
                c.yield(BoundaryStreamToken(index: 3, text: "", isFinal: true, managedThreadId: 9, elapsedUs: 2400))
                c.finish()
            }
        }
    }
}

File: android/app/src/main/kotlin/io/dotnetnativeinterop/boundary/BoundaryService.kt
package io.dotnetnativeinterop.boundary

import io.dotnetnativeinterop.model.BoundaryEcho
import io.dotnetnativeinterop.model.BoundaryStreamToken
import io.dotnetnativeinterop.model.BoundaryThrow
import io.dotnetnativeinterop.model.OwnershipEntry
import io.dotnetnativeinterop.model.PhaseTiming
import io.dotnetnativeinterop.transport.FfiTraceListener
import io.dotnetnativeinterop.transport.NativeBridge
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json

internal val boundaryJson: Json = Json { ignoreUnknownKeys = true }

/** Result of one echo trace: native echo + frontend timing + thread context. */
public data class BoundaryEchoTrace(
    val echo: BoundaryEcho,
    val timing: PhaseTiming,
    val callerThreadId: Long,
)

public interface BoundaryService {
    public suspend fun echo(input: String): BoundaryEchoTrace
    public suspend fun throwDemo(): BoundaryThrow
    public fun stream(prompt: String, maxTokens: Int): Flow<BoundaryStreamToken>
}

/** BoundaryService over the in-process C ABI (Pattern 3 — boundary instrumentation). */
public class FfiBoundaryService : BoundaryService {

    override suspend fun echo(input: String): BoundaryEchoTrace = withContext(Dispatchers.IO) {
        val callerTid = Thread.currentThread().id
        val cStart = System.nanoTime()
        val raw = NativeBridge.nativeFfiEcho(input) ?: error("nativeFfiEcho returned null")
        val callWallUs = (System.nanoTime() - cStart) / 1000.0
        val echo = boundaryJson.decodeFromString<BoundaryEcho>(raw)
        // marshal/free happen inside the JNI shim (not separately observable from Kotlin) — fold into cross;
        // executeUs is native; cross = round-trip - native execute. (Honest split is labelled in the UI.)
        val timing = PhaseTiming(
            crossUs = (callWallUs - echo.executeUs).coerceAtLeast(0.0),
            executeUs = echo.executeUs,
        )
        BoundaryEchoTrace(echo, timing, callerTid)
    }

    override suspend fun throwDemo(): BoundaryThrow = withContext(Dispatchers.IO) {
        val raw = NativeBridge.nativeFfiThrow() ?: error("nativeFfiThrow returned null")
        boundaryJson.decodeFromString(raw)
    }

    override fun stream(prompt: String, maxTokens: Int): Flow<BoundaryStreamToken> = flow {
        val channel = Channel<BoundaryStreamToken>(capacity = 64)
        val listener = object : FfiTraceListener {
            override fun onTrace(index: Int, text: String, managedThreadId: Long, elapsedUs: Long, isFinal: Boolean) {
                channel.trySend(BoundaryStreamToken(index, text, isFinal, managedThreadId, elapsedUs))
                if (isFinal) channel.close()
            }
        }
        val sessionId = withContext(Dispatchers.Default) {
            NativeBridge.nativeFfiStreamStart(prompt, maxTokens, listener)
        }
        if (sessionId <= 0) throw IllegalStateException("nativeFfiStreamStart failed: status $sessionId")
        try {
            for (token in channel) emit(token)
        } finally {
            // Stop via the shared lifecycle exports (no-ops if the stream already finished).
            withContext(Dispatchers.Default) {
                NativeBridge.nativeSessionCancel(sessionId)
                NativeBridge.nativeSessionFree(sessionId)
            }
        }
    }.flowOn(Dispatchers.Default)
}

/** Ownership ledger rows. On Android the JNI shim frees the result string eagerly (take_native_string ->
 *  dni_string_free), so the "leak" is illustrative, not a real leak (unlike iOS where the caller owns it). */
public fun echoLedger(inputBytes: Int, resultBytes: Int): List<OwnershipEntry> = listOf(
    OwnershipEntry("input utf8", "host", "JNI (ReleaseStringUTFChars)", inputBytes, freed = true),
    OwnershipEntry("result json", ".NET", "JNI (dni_string_free)", resultBytes, freed = true),
)

/** Deterministic fake for unit tests + previews. */
public class FakeBoundaryService : BoundaryService {
    override suspend fun echo(input: String): BoundaryEchoTrace {
        val hex = input.encodeToByteArray().joinToString("") { "%02X".format(it) }
        return BoundaryEchoTrace(
            BoundaryEcho(hex, input.encodeToByteArray().size, input, 7L, 4.2, "0x1000"),
            PhaseTiming(crossUs = 2.0, executeUs = 4.2), 1L,
        )
    }
    override suspend fun throwDemo(): BoundaryThrow =
        BoundaryThrow(true, "System.InvalidOperationException", "Boundary demo: contained.", -5)
    override fun stream(prompt: String, maxTokens: Int): Flow<BoundaryStreamToken> = flow {
        repeat(3) { emit(BoundaryStreamToken(it, "tok$it ", false, 9L, it * 800L)) }
        emit(BoundaryStreamToken(3, "", true, 9L, 2400L))
    }
}
